rust/lance/src/io/writer/statistics.rs

```diff
@@ -0,0 +1,1421 @@
+
+/// Max number of bytes that are included in statistics for binary columns.
+const BINARY_PREFIX_LENGTH: usize = 64;
+
+/// Statistics for a single column chunk.
+#[derive(Debug, PartialEq)]
+pub struct StatisticsRow {
+    /// Number of nulls in this column chunk.
+    pub(crate) null_count: ScalarValue,
+    /// Minimum value in this column chunk, if any
+    pub(crate) min_value: ScalarValue,
+    /// Maximum value in this column chunk, if any
+    pub(crate) max_value: ScalarValue,
+}
+
+fn get_statistics<T: ArrowNumericType>(arrays: &[&ArrayRef]) -> StatisticsRow
+where
+    T::Native: Bounded + PartialOrd,
+    datafusion::scalar::ScalarValue: From<<T as ArrowPrimitiveType>::Native>,
+{
+    let arr = arrays
+        .iter()
+        .map(|x| x.as_any().downcast_ref::<PrimitiveArray<T>>().unwrap())
+        .collect::<Vec<_>>();
+    let mut min_value = T::Native::max_value();
+    let mut max_value = T::Native::min_value();
+    let mut null_count: i64 = 0;
+
+    for array in arr.iter() {
+        array.iter().for_each(|value| {
+            if let Some(value) = value {
+                if let Some(Ordering::Greater) = value.partial_cmp(&max_value) {
+                    max_value = value;
+                } else if let Some(Ordering::Less) = value.partial_cmp(&min_value) {
+                    min_value = value;
+                }
+            };
+        });
+        null_count += array.null_count() as i64;
+    }
+
+    let mut scalar_min_value = ScalarValue::try_from(min_value).unwrap();
+    let mut scalar_max_value = ScalarValue::try_from(max_value).unwrap();
+
+    let dt = arrays[0].data_type();
+    match dt {
+        DataType::Float32 => {
+            if scalar_min_value == ScalarValue::Float32(Some(0.0)) {
+                scalar_min_value = ScalarValue::Float32(Some(-0.0));
+            }
+            if scalar_max_value == ScalarValue::Float32(Some(-0.0)) {
+                scalar_max_value = ScalarValue::Float32(Some(0.0));
+            }
+        }
+        DataType::Float64 => {
+            if scalar_min_value == ScalarValue::Float64(Some(0.0)) {
+                scalar_min_value = ScalarValue::Float64(Some(-0.0));
+            }
+            if scalar_max_value == ScalarValue::Float64(Some(-0.0)) {
+                scalar_max_value = ScalarValue::Float64(Some(0.0));
+            }
+        }
+        DataType::Date32 => {
+            let min_value_date32 = scalar_min_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int32Array>()
+                .unwrap()
+                .value(0);
+            let max_value_date32 = scalar_max_value
+                .to_array()
+                .as_any()
+                .downcast_ref::<Int32Array>()
+                .unwrap()
+                .value(0);
+            scalar_min_value = ScalarValue::Date32(Some(min_value_date32));
+            scalar_max_value = ScalarValue::Date32(Some(max_value_date32));
+        }
```

Comment:
`min_value` and `max_value` are already the correct type, right? I don't think you need the intermediary scalar and you definitely don't need to convert each one to an array to downcast.
```suggestion
        DataType::Date32 => {
            scalar_min_value = ScalarValue::Date32(Some(min_value));
            scalar_max_value = ScalarValue::Date32(Some(max_value));
        }
```

Reply:
This causes similar problem than described here: https://github.com/lancedb/lance/pull/1335#discussion_r1349370504.
I might want to rethink how I get native types from arrow ones.

Reply:
I think this is now refactored out.